tests/src/Kernel/EntityControllerCacheTest.php

```diff
@@ -0,0 +1,238 @@
+<?php
+
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License version 2 as published by the
+ * Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+namespace Drupal\Tests\apigee_edge\Kernel;
+
+use Apigee\Edge\Api\Management\Entity\CompanyApp;
+use Apigee\Edge\Api\Management\Entity\Developer;
+use Apigee\Edge\Api\Management\Entity\DeveloperApp;
+use Drupal\apigee_edge\Entity\Controller\Cache\EntityCacheInterface;
+use Drupal\apigee_edge\Entity\Controller\Cache\EntityIdCacheInterface;
+use Drupal\KernelTests\KernelTestBase;
+
+/**
+ * Apigee Edge entity controller cache tests.
+ *
+ * @group apigee_edge
+ * @group apigee_edge_kernel
+ */
+class EntityControllerCacheTest extends KernelTestBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected static $modules = [
+    'system',
+    'apigee_edge',
+    'key',
+  ];
+
+  /**
+   * Tests developer entity controller cache.
+   */
+  public function testDeveloperEntityControllerCache() {
+    $developer_cache = $this->container->get('apigee_edge.controller.cache.developer');
+    $developer_id_cache = $this->container->get('apigee_edge.controller.cache.developer_ids');
+
+    // Generate developer entities with random data.
+    $developers = [];
+    for ($i = 0; $i < 3; $i++) {
+      $id = $this->getRandomUniqueId();
+      $developers[$id] = new Developer([
+        'developerId' => $id,
+        'email' => strtolower($this->randomMachineName()) . '@example.com',
+      ]);
+    }
+
+    $this->saveAllEntitiesAndValidate($developers, $developer_cache, $developer_id_cache);
+
+    /** @var \Apigee\Edge\Api\Management\Entity\DeveloperInterface $developer */
+    foreach ($developers as $developer) {
+      // Load developer by email and by id.
+      $this->assertSame($developer, $developer_cache->getEntity($developer->getEmail()));
+      $this->assertSame($developer, $developer_cache->getEntity($developer->getDeveloperId()));
+      $this->assertContains($developer->getEmail(), $developer_id_cache->getIds());
+
+      // Remove developer from cache. Use an invalid id to test whether the
+      // removeEntities() function properly removes the invalid entity IDs.
```

Comment:
"properly removes"? :)
Rather "Pass an invalid entity id to ensure its does not cause any trouble. `@see` \Drupal\apigee_edge\Entity\Controller\Cache\EntityCache::removeEntities() `@see` https://www.drupal.org/project/drupal/issues/3017753"